raven/processes/wps_nalcms_zonal_stats.py

```diff
@@ -0,0 +1,134 @@
+                         data_type='boolean', default='true',
+                         min_occurs=1, max_occurs=1),
+            LiteralInput('select_all_touching', 'Additionally select boundary pixels that are touched by shape',
+                         data_type='boolean', default='false',
+                         min_occurs=1, max_occurs=1),
+        ]
+
+        outputs = [
+            ComplexOutput('statistics', 'DEM properties within the region defined by `shape`.',
+                          abstract='Elevation statistics: min, max, mean, median, sum, nodata',
+                          supported_formats=[FORMATS.JSON, FORMATS.GEOJSON]),
+        ]
+
+        super(NALCMSZonalStatisticsProcess, self).__init__(
+            self._handler,
+            identifier="nalcms-zonal-stats",
+            title="NALCMS Land Use Zonal Statistics",
+            version="1.0",
+            abstract="Return zonal statistics for the CEC NALCMS based on the boundaries of a vector file.",
+            metadata=[],
+            inputs=inputs,
+            outputs=outputs,
+            status_supported=True,
+            store_supported=True)
+
+    def _handler(self, request, response):
+
+        # TODO: Deploy the CEC NALCMS 2010 data set on PAVICS GeoServer and access via WCS
+        raster_url = request.inputs['raster'][0].file
+
+        shape_url = request.inputs['shape'][0].file
+        geojson_out = request.inputs['return_geojson'][0].data
+        touches = request.inputs['select_all_touching'][0].data
+
+        vectors = ['.gml', '.shp', '.gpkg', '.geojson', '.json']
+        vector_file = single_file_check(archive_sniffer(shape_url, working_dir=self.workdir, extensions=vectors))
+        rasters = ['.tiff', '.tif']
+        raster_file = single_file_check(archive_sniffer(raster_url, working_dir=self.workdir, extensions=rasters))
+
+        vec_crs, ras_crs = crs_sniffer(vector_file), crs_sniffer(raster_file)
+
+        if vec_crs != ras_crs:
+            msg = 'CRS for files {} and {} are not the same. Reprojecting...'.format(vector_file, raster_file)
+            LOGGER.warning(msg)
+
+            # Reproject full vector to preserve feature attributes
+            projected = tempfile.NamedTemporaryFile(prefix='reprojected_', suffix='.json', delete=False,
+                                                    dir=self.workdir).name
+            generic_vector_reproject(vector_file, projected, driver='GeoJSON', source_crs=vec_crs, target_crs=ras_crs)
+            vector_file = projected
+
+        try:
+            stats = zonal_stats(
+                vector_file, raster_file, stats=['count', 'nodata'],
+                band=1, categorical=True, category_map=NALCMS_CATEGORIES, all_touched=touches,
+                geojson_out=geojson_out, raster_out=False)
+
+            if not geojson_out:
+                response.outputs['statistics'].data = json.dumps(stats)
+
+            else:
+                if len(stats) > 1:
```

Comment:
It's really a pain on the client side when the outputs change their topology depending on input parameters. In other words, we expect that outputs are a function of the inputs, but the container of the outputs should not change. Here what does the FeatureCollection add ? I suggest outputs are either always FeatureCollections (with one item sometimes) or never FeatureCollections.

Reply:
FeatureCollections are the GeoJSON standard for multi-feature vector files. I can set things so that they always output FeatureCollections regardless of number of features and there shouldn't be any problems.